0000-clear_on_drop.md

```diff
@@ -0,0 +1,225 @@
+- Feature Name: `clear_on_drop`
+- Start Date: 2016-02-10
+- RFC PR:
+- Rust Issue:
+
+# Summary
+[summary]: #summary
+
+This RFC proposes a pair of attributes, `#[clear_on_drop]` and
+`#[clear_stack_on_return]`, to make it easy to write code which securely
+clears sensitive data (for instance, encryption keys) after its use.
+
+# Motivation
+[motivation]: #motivation
+
+Some kinds of data should not be kept in memory any longer than they are
+needed. For instance, the TLS protocol has Ephemeral Diffie-Hellman
+modes which give a connection the forward secrecy property — but only if
+both the ephemeral key and the session keys are securely discarded after
+the connection is over.
+
+For long-lived processes, unless carefully overwritten, pieces of
+sensitive data can linger in the heap, the stack, and even in the
+processor registers. Clearing the heap and explicitly allocated
+variables in the stack is somewhat easy, needing only careful attention
+to not forget any exit path, and some way to defeat the compiler's dead
+store optimizations. Clearing compiler-created temporary variables in
+the stack and processor registers is harder.
+
+This RFC proposes a pair of attributes to simplify this task, and reduce
+the chance of mistakes. The first one, `#[clear_on_drop]`, asks the
+compiler to, after any drop or move of a variable of the given type,
+securely overwrite its contents. The second one,
+`#[clear_stack_on_return]`, asks the compiler to, just before a given
+function returns, securely overwrite all the stack space and processor
+registers it used.
+
+(In this text, "securely overwrite" means: overwrite with some value
+unrelated to any of the values being securely overwritten, and in such a
+way that no compiler or linker optimization pass will remove the
+overwrite, without also removing the original write(s). The exact value
+being used doesn't matter; for instance, a callee-saved register can be
+securely overwritten by the normal restore of the caller's value from
+the stack).
+
+# Detailed design
+[design]: #detailed-design
+
+This RFC proposes two independent attributes, which complement each
+other. They will be explained separately.
+
+## `#[clear_on_drop]`
+
+The `#[clear_on_drop]` attribute can be attached to a `struct` or to an
+`enum`. Conceptually, the attribute modifies the type's `Drop` impl
+(adding one if needed) so that, after it finishes:
+
+* The memory used by the variable is securely overwritten;
+* For non-`Copy` types, the memory used by it before each move is
+  securely overwritten.
+
+Clearly, it's impractical to track the memory used before every move, to
+overwrite it all in the end; therefore, in practice `#[clear_on_drop]`
+also asks the compiler to securely overwrite the old copy after every
+move (for `Copy` types, the old copy will be overwritten by its `Drop`).
```

Comment:
The error message links to https://github.com/rust-lang/rust/issues/20126, which says implementing `Copy` and `Drop` together is currently disabled because the implementation was broken.

Of course, that would either have to be fixed to be able to implement `#[clear_on_drop]` on `Copy` types, or `#[clear_on_drop]` would have to initially be restricted to non-`Copy` types. I believe being unable to use it with `Copy` would not be a big restriction.